include/rmm/cuda_stream_view.hpp

```diff
@@ -0,0 +1,154 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <rmm/detail/error.hpp>
+
+#include <cuda_runtime_api.h>
+
+#include <atomic>
+#include <cstddef>
+#include <cstdint>
+
+namespace rmm {
+
+/**
+ * @brief Strongly-typed non-owning wrapper for CUDA streams with default constructor.
+ *
+ * This wrapper is simply a "view": it does not own the lifetime of the stream it wraps.
+ */
+class cuda_stream_view {
+ public:
+  constexpr cuda_stream_view()                        = default;
+  constexpr cuda_stream_view(cuda_stream_view const&) = default;
+  constexpr cuda_stream_view(cuda_stream_view&&)      = default;
+  constexpr cuda_stream_view& operator=(cuda_stream_view const&) = default;
+  constexpr cuda_stream_view& operator=(cuda_stream_view&&) = default;
+  ~cuda_stream_view()                                       = default;
+
+  // TODO disable construction from 0 after cuDF and others adopt cuda_stream_view
+  // cuda_stream_view(int)            = delete; //< Prevent cast from 0
+  // cuda_stream_view(std::nullptr_t) = delete; //< Prevent cast from nullptr
+  // TODO also disable implicit conversion from cudaStream_t
+
+  /**
+   * @brief Implicit conversion from cudaStream_t.
+   */
+  constexpr cuda_stream_view(cudaStream_t stream) noexcept : stream_{stream} {}
+
+  /**
+   * @brief Get the wrapped stream.
+   *
+   * @return cudaStream_t The wrapped stream.
+   */
+  constexpr cudaStream_t value() const noexcept { return stream_; }
+
+  /**
+   * @brief Explicit conversion to cudaStream_t.
+   */
+  explicit constexpr operator cudaStream_t() const noexcept { return value(); }
+
+  /**
+   * @brief Explicit conversion to uintptr_t.
+   */
+  explicit operator uintptr_t() const noexcept { return reinterpret_cast<uintptr_t>(value()); }
+
+  /**
+   * @brief Return true if the wrapped stream is the CUDA default stream (aka Null Stream).
+   */
+  constexpr bool is_default() const noexcept { return value() == cudaStreamDefault; }
+
+  /**
+   * @brief Return true if the wrapped stream is the CUDA per-thread default stream.
+   */
+  bool is_per_thread_default() const noexcept
+  {
+#ifdef CUDA_API_PER_THREAD_DEFAULT_STREAM
+    return value() == cudaStreamPerThread || value() == cudaStreamDefault;
+#else
+    return value() == cudaStreamPerThread;
+#endif
+  }
+
+  /**
+   * @brief Return true if the wrapped stream is explicitly the CUDA legacy default stream.
+   */
+  bool is_legacy_default() const noexcept { return value() == cudaStreamLegacy; }
```

Comment:
If I'm understanding the docs correctly (and I very well may not be), then `CUDA_API_PER_THREAD_DEFAULT_STREAM` effectively toggles the behavior of `cudaStreamDefault`, such that it behaves either as cudaPerStreamStream, _or_ the legacy stream. If that's the case, what if we forgo the mention of "default" altogether, and instead refer to the streams as either "legacy" or "per thread"? (we may also want to add a special method for `cudaStreamNonBlocking`).
```suggestion
  bool is_per_thread_stream() const noexcept
  {
#ifdef CUDA_API_PER_THREAD_DEFAULT_STREAM
    return value() == cudaStreamPerThread || value() == cudaStreamDefault;
#else
    return value() == cudaStreamPerThread;
#endif
  }

  /**
   * @brief Return true if the wrapped stream is explicitly the CUDA legacy default stream.
   */
  bool is_legacy_stream() const noexcept {  
#ifdef CUDA_API_PER_THREAD_DEFAULT_STREAM
    return value() == cudaStreamLegacy;
#else
    return value() == cudaStreamLegacy || value() == cudaStreamDefault;
#endif
}
```

Reply:
To clarify, I think it would be helpful to have functions that indicate the _behavior_ of the stream. The fact that the stream is `cudaStreamDefault` is maybe beside the point, and therefore mention of "default" might just be confusing.

Reply:
I'm torn. While this looks nice, I'm trying to define a type-safe wrapper around CUDA semantics, not define new semantics. So I think having the naming reflect CUDA naming is the least ambiguous.  On the other hand I already made `is_per_thread_default()` do more than just compare to `cudaStreamPerThread`, and this has proved its utility in two places in the library already...

(I would not drop the word "default" from the function names in either case, because I think "per-thread stream" and "per-thread default stream" have potentially different meanings, and similar for `legacy_default` vs. `legacy`.)

Reply:
Decided to do what you suggest but call it `is_default` and drop `is_legacy_default`.

Reply:
If we don't want to introduce new semantics, maybe we can give the caller the option?
```C++

#ifdef CUDA_API_PER_THREAD_DEFAULT_STREAM
constexpr bool is_ptds_enabled = true;
#else
constexpr bool is_ptds_enabled = false;
#endif

bool is_cuda_per_thread_stream(bool detect_default = false)
{
    if (value() == cudaStreamPerThread) {
      return true;
    }

    if (detect_default and is_ptds_enabled) 
    {
      return value() == cudaStreamDefault;
    }

    return false;
}

bool is_cuda_legacy_stream(bool detect_default = false)
{
    if (value() == cudaStreamLegacy) {
      return true;
    }

    if (detect_default and not is_ptds_enabled) {
      return value() == cudaStreamDefault;
    }

    return false;
}

bool is_cuda_default_stream()
{
    return value() == cudaStreamDefault;
}

bool is_cuda_non_blocking_stream()
{
    return value() == cudaStreamNonBlocking;
}
```

Reply:
I think the user has everything they need via `value()` and these two functions.